cli-commands/verifier/verifier.js

```diff
@@ -0,0 +1,162 @@
+const { runWithoutWriteFiles } = require('../flattener/flatten');
+const axios = require('axios');
+const querystring = require('querystring');
+const logger = require('../../logger-service/logger-service').logger;
+const colors = require('../../utils/colors');
+const ethers = require('ethers');
+const DEFAULT_SEND_REQUEST_TIMEOUT = 10000;
+const DEFAULT_CHECK_STATUS_TIMEOUT = 5000;
+const MODULE_NAME = 'contract';
+const ACTION_VERIFY_SOURCE = 'verifysourcecode';
+const ACTION_VERIFY_STATUS = 'checkverifystatus';
+class Verifier {
+
+	constructor() {
+
+	}
+
+	async verifySmartContract(contractWrapper, deploymentArguments, libraries, defaultOverrides) {
+
+		const etherscanApiKey = defaultOverrides.etherscanApiKey;
+		const contractName = contractWrapper._contract.contractName;
+		const flattenedCode = await this._flattenSourceCode(contractWrapper);
+		const constructorArguments = await this._buildConstructorArguments(contractWrapper, deploymentArguments);
+		const contractLibraries = this._buildLibrariesArguments(libraries);
+		const { apiUrl } = await this._buildApiUrl(contractWrapper);
+		const { networkName } = await this._buildApiUrl(contractWrapper);
+		const solcVersionCompiler = this._buildSolcVersionCompiler(contractWrapper._contract.compiler.version);
+		logger.log(`Attempting to verify your contract: ${colors.colorName(contractName)} on network ${colors.colorParams(networkName)}`);
+		let data = this._constructRequestData(etherscanApiKey, contractWrapper, contractLibraries, flattenedCode, solcVersionCompiler, constructorArguments);
+
+		const response = await this._sendVerificationRequest(data, defaultOverrides, apiUrl);
+		return await this._checkVerificationStatus(response, defaultOverrides, contractName, apiUrl);
+	}
+
+	async _flattenSourceCode(contractWrapper) {
+		const regexp = /[0-9]?\.[0-9]?\.[0-9]?/;
+		const solcVersion = regexp.exec(contractWrapper._contract.compiler.version)[0];
+		const sourceCode = await runWithoutWriteFiles(`${contractWrapper._contract.contractName}.sol`, solcVersion);
+		return sourceCode
+	}
+
+	async _buildConstructorArguments(contractWrapper, deploymentArguments) {
+		if (!deploymentArguments.length) {
+			return;
+		}
+		let encoder = new ethers.utils.AbiCoder();
+		let types = [];
+		contractWrapper.interface.deployFunction.inputs.forEach(element => {
```

Comment:
in some implementations of nodeJS forEach changes the order of elements. This will break your logic as you need the deploy function input types to be in the correct order. Please use standard `for`